### ISSUE TITLE
Memory consumption experiments?

### ISSUE THREAD

**Issue by wallykop** [Yura52/tabular-dl-revisiting-models#23]
Hello! 

I would like to ask you, whether you've done the experiments on the memory consumption of the elaborated models and its comparison with each other?

**Comment by Yura52** [comment]
Hello!

We did not run measure memory consumption. As for such "technical" metrics, for many experiments, we report model sizes (as the number of trainable weights), training time and the number of the used GPUs. For example, [in this report](https://github.com/Yura52/tabular-dl-revisiting-models/blob/f65a93329a58c80ea486a212feeba6babf9fdd4f/output/california_housing/mlp/tuned/0/stats.json), you can find these metrics under the keys "n_parameters", "time" and "CUDA_VISIBLE_DEVICES", respectively.

**Comment by wallykop** [comment]
Thank you for your feedback! 

However, I think, that estimating memory consumption using number of trainable weights, training time and number of GPUs might give misleading insights. 

Originally, my question arose with my attempts of increasing batch_size for the FT model. I trained ResNet and FT on the GPUs, however, for the FT there have been a lack of allocated resources. 

Therefore, I decided to compare the memory consumption on my data of the forward pass on the GPU between ResNet (2349001 trainable weights) and FT (3850255 trainable weights). For the batch_size = 2, the results are the following (calculated using torch.cuda.memory_allocated(device) before and after forward pass): 

* ResNet: Memory consumed by forward pass 90624 bytes
  
* FT: Memory consumed by forward pass 14984192 bytes

The factor of difference is nearly 160 times. 

Therefore, I would like to ask you, whether you have had similar memory consumption on your datasets and whether there exists methods to reduce memory usage of the FT Transformer?

**Comment by Yura52** [comment]
> However, I think, that estimating memory consumption using number of trainable weights, training time and number of GPUs might give misleading insights. 

Yes, I mentioned those metrics just in the case if they can be useful for you as well :)

Anyway, FT-Transformer is expected to consume much more memory.

- on our datasets, using the `train_with_auto_virtual_batch` function ([example](https://github.com/Yura52/tabular-dl-revisiting-models/blob/f65a93329a58c80ea486a212feeba6babf9fdd4f/bin/ft_transformer.py#L493)) usually helped. This function automatically selects the virtual batch size that fits into memory and performs gradient accumulation.
- if your dataset has many features, you can try using the [linformer options](https://github.com/Yura52/tabular-dl-revisiting-models/blob/f65a93329a58c80ea486a212feeba6babf9fdd4f/bin/ft_transformer.py#L178). Example: [link](https://github.com/Yura52/tabular-dl-revisiting-models/blob/f65a93329a58c80ea486a212feeba6babf9fdd4f/output/epsilon/ft_transformer/tuned/0.toml#L13).
- you can copy [this](https://github.com/Yura52/zero/blob/877e3abb40259e377199bc202d14ddcafe949984/zero/hardware.py#L11) function and call it as often as possible just to check if it helps at all.
- also, you can try move your code under `main()` function, because, sometimes, when everything is in the global namespace, it is harder for Python to delete unnecessary objects and free memory

**Comment by wallykop** [comment]
Thank you for your ideas!
